Mathlib/RingTheory/PowerSeries/WeierstrassPreparation.lean

```diff
@@ -0,0 +1,617 @@
+its maximal ideal. -/
+abbrev IsWeierstrassDivision [IsLocalRing A] : Prop :=
+  f.IsWeierstrassDivisionAt g q r (IsLocalRing.maximalIdeal A)
+
+theorem isWeierstrassDivisionAt_zero : IsWeierstrassDivisionAt 0 g 0 0 I := by
+  constructor
+  · rw [Polynomial.degree_zero]
+    exact WithBot.bot_lt_coe _
+  · simp
+
+variable {f g q r I}
+
+theorem IsWeierstrassDivisionAt.coeff_f_sub_r_mem (H : f.IsWeierstrassDivisionAt g q r I) :
+    ∀ i < (g.map (Ideal.Quotient.mk I)).order.toNat, coeff A i (f - r) ∈ I := fun i hi ↦ by
+  replace H := H.2
+  rw [← sub_eq_iff_eq_add] at H
+  rw [H]
+  refine coeff_mul_mem_ideal_of_coeff_left_mem_ideal i (fun j hj ↦ ?_) i le_rfl
+  have := coeff_of_lt_order_toNat _ (lt_of_le_of_lt hj hi)
+  rwa [coeff_map, ← RingHom.mem_ker, Ideal.mk_ker] at this
+
+end IsWeierstrassDivisionAt
+
+section IsWeierstrassDivisorAt
+
+variable (g : A⟦X⟧) (I : Ideal A)
+
+/-- `PowerSeries.IsWeierstrassDivisorAt g I` is a `Prop` which asserts that the `n`-th coefficient
+of `g` is a unit, where `n` is the order of the
+image of `g` in `(A / I)⟦X⟧` (defined to be zero if such image is zero, in which case
+it's mathematically not considered).
+
+This property guarantees that if the ring is `I`-adic complete, then `g` can be used as a divisor
+in Weierstrass division (`PowerSeries.IsWeierstrassDivisorAt.isWeierstrassDivisionAt_div_mod`). -/
+def IsWeierstrassDivisorAt : Prop :=
+  IsUnit (coeff A (g.map (Ideal.Quotient.mk I)).order.toNat g)
+
+/-- Version of `PowerSeries.IsWeierstrassDivisorAt` for local rings with respect to
+its maximal ideal. -/
+abbrev IsWeierstrassDivisor [IsLocalRing A] : Prop :=
+  g.IsWeierstrassDivisorAt (IsLocalRing.maximalIdeal A)
+
+/-- If `g` is a power series over a local ring such that
+its image in the residue field is not zero, then `g` can be used as a Weierstrass divisor. -/
+theorem IsWeierstrassDivisor.of_map_ne_zero [IsLocalRing A]
+    (hg : g.map (IsLocalRing.residue A) ≠ 0) : g.IsWeierstrassDivisor := by
+  rw [IsWeierstrassDivisor, IsWeierstrassDivisorAt, ← IsLocalRing.not_mem_maximalIdeal]
+  have h := coeff_order hg
+  contrapose! h
+  rwa [coeff_map, IsLocalRing.residue_eq_zero_iff]
+
+private theorem coeff_trunc_order_mem (i : ℕ) :
+    (g.trunc (g.map (Ideal.Quotient.mk I)).order.toNat).coeff i ∈ I := by
+  rw [coeff_trunc]
+  split_ifs with h
+  · simpa only [coeff_map, ← RingHom.mem_ker, Ideal.mk_ker] using coeff_of_lt_order_toNat _ h
+  · exact zero_mem _
+
+namespace IsWeierstrassDivisorAt
+
+variable {g I} (H : g.IsWeierstrassDivisorAt I)
+include H
+
+theorem isUnit_shift : IsUnit <| PowerSeries.mk fun i ↦
+    coeff A (i + (g.map (Ideal.Quotient.mk I)).order.toNat) g := by
+  simpa only [isUnit_iff_constantCoeff, constantCoeff_mk, zero_add]
+
+/-- The inductively constructed sequence `qₖ` in the proof of Weierstrass division. -/
+noncomputable def seq (H : g.IsWeierstrassDivisorAt I) (f : A⟦X⟧) : ℕ → A⟦X⟧
```

Comment:
I think this should be `private`.

Reply:
I want to expose this construction here in case of someone want to use it. Otherwise the whole `IsWeierstrassDivisorAt` should be private.

Reply:
OK, no problem.